Barrier.cpp fails compilation with ACE_HAS_DUMP defined in Visual Studio 2022.

All support on github is best effort only, no guarantees given. When you need guaranteed support have a look at the commercial support offerings as listed at https://github.com/DOCGroup/ACE_TAO/wiki/ACE-and-TAO-Commercial-support.

### Version

The version of ACE and/or TAO you are using:  
7.1.3, also present in latest github.

### Host machine and operating system
Windows 11 23H2

### Target machine and operating system (if different from host)


### Compiler name and version (including patch level)
Visual Studio 2022  17.12.1

### The $ACE_ROOT/ace/config.h file
#ifdef _DEBUG
#define ACE_HAS_DUMP 1
#endif
#define TAO_HAS_ZIOP 1
#ifndef wxHAS_MODE_T
#define  wxHAS_MODE_T 1
#endif
#ifndef _WIN32_WINNT
#define _WIN32_WINNT _WIN32_WINNT_WIN7
#define NTDDI_VERSION NTDDI_WIN7
#endif
#ifndef _WINSOCK_DEPRECATED_NO_WARNINGS
#define _WINSOCK_DEPRECATED_NO_WARNINGS
#endif
#include "ace/config-win32.h"

If you use a link to a platform-specific file, simply state which one

### The $ACE_ROOT/include/makeinclude/platform_macros.GNU file

if you use a link to a platform-specific file, simply state which one (unless this isn't used in this case, e.g., with Microsoft Visual C++)

### Contents of $ACE_ROOT/bin/MakeProjectCreator/config/default.features

Used by MPC when you generate your own makefiles

### AREA/CLASS/EXAMPLE AFFECTED:

What example failed?  What module failed to compile?
ace/Barrier.cpp 


### The problem effects:

Compilation with ACE_HAS_DUMP defined

### Synopsis

Error:
  H:\dev-vc14\3rdParty\ACE+TAO-7.1.3-VC16\ace\Barrier.cpp(28,3): error C4839: non-standard use of class 'std::atomic<int>' as an argument to a variadic function [H:\dev-vc14\3rdParty\ACE+TAO-7.1.3-VC16\ace\ACE_vs2019.vcxproj]
         H:\dev-vc14\3rdParty\ACE+TAO-7.1.3-VC16\ace\Barrier.cpp(28,3): error C2280: 'std::atomic<int>::atomic(const std::atomic<int> &)': attempting to reference a deleted function [H:\dev-vc14\3rdParty\ACE+TAO-7.1.3-VC16\ace\ACE_vs2019.vcxproj]


### Description

Problem line
<snip line 28>
ACELIB_DEBUG ((LM_DEBUG, ACE_TEXT ("running_threads_ = %d\n"), this->running_threads_));
</sniip>

running_threads_ was converted from an int to a std::atomic<int> between 7.1.2 and 7.1.3.  Atomic int's can't be passed to a variadic function without a cast.


### Repeat by

Defining ACE_HAS_DUMP in config.h

### Sample fix/ workaround

ACELIB_DEBUG ((LM_DEBUG, ACE_TEXT ("running_threads_ = %d\n"), static_cast<int>(this->running_threads_)));

--- a/ACE/ace/Barrier.cpp
+++ b/ACE/ace/Barrier.cpp
@@ -25,7 +25,7 @@ ACE_Sub_Barrier::dump () const
 
   ACELIB_DEBUG ((LM_DEBUG, ACE_BEGIN_DUMP, this));
   this->barrier_finished_.dump ();
-  ACELIB_DEBUG ((LM_DEBUG, ACE_TEXT ("running_threads_ = %d\n"), this->running_threads_));
+  ACELIB_DEBUG ((LM_DEBUG, ACE_TEXT ("running_threads_ = %d\n"), this->running_threads_.load()));
   ACELIB_DEBUG ((LM_DEBUG, ACE_END_DUMP));
 #endif /* ACE_HAS_DUMP */
 }

